kernel/signal.c

```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (c) 2023, Meta
+ * Copyright (c) 2024, Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdalign.h>
+
+#include <zephyr/internal/syscall_handler.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/math_extras.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#ifndef BITS_PER_BYTE
+#define BITS_PER_BYTE 8
+#endif
+
+#define K_NSIG CONFIG_SIGNAL_SET_SIZE
+
+LOG_MODULE_REGISTER(k_sig, CONFIG_SIGNAL_LOG_LEVEL);
+
+struct k_sig_fifo_element {
+	k_tid_t tid;
+	union k_sig_val value;
+	int signo;
+};
+
+/*
+ * We use a ring buffer of bytes, where each byte is an index of an allocated k_sig_fifo_element
+ * from the slab below. This should save some space, since 256 signals should be well beyond
+ * enough to queue.
+ */
+BUILD_ASSERT(CONFIG_SIGNAL_QUEUE_SIZE > 0, "SIGQUEUE_MAX is too small");
+BUILD_ASSERT(CONFIG_SIGNAL_QUEUE_SIZE <= BIT(BITS_PER_BYTE), "SIGQUEUE_MAX is too large");
+
+BUILD_ASSERT(CONFIG_SIGNAL_SET_SIZE >= 0);
+
+RING_BUF_DECLARE_STATIC(k_sig_fifo, CONFIG_SIGNAL_QUEUE_SIZE);
+static struct k_poll_signal k_sig_fifo_ready = K_POLL_SIGNAL_INITIALIZER(k_sig_fifo_ready);
+K_MEM_SLAB_DEFINE_STATIC(k_sig_slab, sizeof(struct k_sig_fifo_element), CONFIG_SIGNAL_QUEUE_SIZE,
+			 alignof(struct k_sig_fifo_element));
+static struct k_sig_fifo_element *const k_sig_array =
+	(struct k_sig_fifo_element *)_k_mem_slab_buf_k_sig_slab;
+static struct k_spinlock *const k_sig_lock = &k_sig_slab.lock;
+
+static int k_sig_min_rt_in_set(const struct k_sig_set *set);
+static bool k_sig_setisempty(const struct k_sig_set *set);
+static void k_sig_unset(struct k_sig_set *result, const struct k_sig_set *set);
+static void k_sig_orset(struct k_sig_set *result, const struct k_sig_set *set2);
+static void k_sig_fifo_dump(void);
+static inline void k_sig_dump_set(const struct k_sig_set *set, const char *label);
+
+/** @brief True if thread id @a tid is valid */
+static bool k_sig_pid_is_valid(k_pid_t pid)
+{
+	k_tid_t tid = (k_tid_t)pid;
+	const uint32_t valid_states = _THREAD_PENDING | _THREAD_SUSPENDED | _THREAD_QUEUED;
+
+	if (pid == NULL) {
+		LOG_DBG("pid is NULL");
+		return false;
+	}
+
+	if (_current == tid) {
+		/* current tid is always valid (thread signals itself) */
+		return true;
+	}
+
+	if ((tid->base.thread_state & valid_states) != 0) {
+		return true;
+	}
+
+	LOG_DBG("thread state %x is invalid (valid mask: %x)", tid->base.thread_state,
+		valid_states);
+
+	return false;
+}
+
+static void k_sig_pending_unlocked(struct k_sig_set *set)
+{
+	struct k_thread *tid = _current;
```

Comment:
What is the best call to make to determine the thread ID associated with a syscall while the syscall is executing? `k_current_get()`?

Reply:
When it comes to getting the current thread ID, __current_ can generally be used by kernel internals--I usually think of k_current_get() as being more application oriented.

I may have misread something, but it does look to me like this routine is ISR callable. It can be called by z_impl_k_sig_queue(), which in turn has documentation that explicitly states that it is safe to call in an ISR context. Given how it gets used, I suspect that all we would need is to make the k_sig_addset() also conditional upon the context not being an ISR.

```
if ((entry->tid == tid) && !arch_is_in_isr()) {
    k_sig_addset(set, entry->signo);
}
```

Reply:
It might be possible to call this with tid as a parameter, which, I think would solidify the use.

It is static here, so use of this function should be from behind a syscall, and not an ISR.